### PR TITLE
Separated Bonk functionality and component from ClimbSystem and ClimbComponent

### PR DESCRIPTION
**Pull request by Telyonok**
## About the PR

Separated BonkableComponent from ClimbableComponent. Changed ClimbSystem and prototypes accordingly. Didn't touch the actual Bonk mechanic.

**Media**

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

### CODE REVIEW

File: Content.Server/Climbing/ClimbSystem.cs [hunk 1]
```diff
@@ -92,7 +92,7 @@ private void AddClimbableVerb(EntityUid uid, ClimbableComponent component, GetVe
         if (!args.CanAccess || !args.CanInteract || !_actionBlockerSystem.CanMove(args.User))
             return;
 
-        if (component.Bonk && _cfg.GetCVar(CCVars.GameTableBonk))
+        if (HasComp<BonkableComponent>(args.Target) && _cfg.GetCVar(CCVars.GameTableBonk))
             return;
```
Comment (AJCM-git) [inline Content.Server/Climbing/ClimbSystem.cs:96]: This isnt your fault but remove this check, it doesnt display the verb when bonking is allowed which makes no sense

**Review by AJCM-git — CHANGES_REQUESTED** [review]
Bonkable is still completely dependent on climbable, to completely decouple it you should make a Bonkable system that has `TryBonk(EntityUid uid, BonkableComponent component)` as a public method, subscribes to `DragDropEvent` and gets called on `ClimbableSystem.TryMoveEntity`  as a dependency. If you do this i recommend you try fix the bug in master that when you drag another entity onto a table, you are the one that gets bonked, i wont require you since it isnt your bug but try give it a shot since you are on this

**Comment by Telyonok** [comment]
> Bonkable is still completely dependent on climbable, to completely decouple it you should make a Bonkable system that has `TryBonk(EntityUid uid, BonkableComponent component)` as a public method, subscribes to `DragDropEvent` and gets called on `ClimbableSystem.TryMoveEntity` as a dependency. If you do this i recommend you try fix the bug in master that when you drag another entity onto a table, you are the one that gets bonked, i wont require you since it isnt your bug but try give it a shot since you are on this

I don't quite understand what for should the BonkSystem subscribe to DragDropEvent?

TryMoveEntity calls TryBonk(EntityUid uid, BonkableComponent component, ClimbableComponent climbableComponent) that returns true or false. And that seems to be working fine.

**Comment by AJCM-git** [comment]
The point of the issue is completely decoupling bonkable from climbable, so that you can add bonkable to an entity without climbable and have the expected behavior, for that bonkable needs to separate its behaviors into a new system, one of them is that there is a chance to bonk someone when they drag drop on a bonkable entity, thus the subscription to the event

**Comment by Telyonok** [comment]
Alright, so TryBonk should react on every DragDrop event on a bonkable entity.

So why should it be called by ClimbSystem.TryMoveEntity at all? I mean it is called twice this way: after drag-and-dropping and by ClimbSystem.TryMoveEntity. I guess I am to just remove everything Bonk related from ClimbSystem. Or am I missing something? Thanks for directions by the way.

**Comment by AJCM-git** [comment]
> So why should it be called by ClimbSystem.TryMoveEntity at all? I mean it is called twice this way: after drag-and-dropping and by ClimbSystem.TryMoveEntity. I guess I am to just remove everything Bonk related from ClimbSystem. Or am I missing something? Thanks for directions by the way.

Because the verb calls `TryMoveEntity`

**Comment by Telyonok** [comment]
Alright, so I think I got it now, yet I am unsure if my solution with adding bool dragDrop to TryMoveEntity is good.

File: Content.Shared/Climbing/BonkableComponent.cs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+using Robust.Shared.GameStates;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Content.Shared.Climbing
+{
+    [RegisterComponent, NetworkedComponent]
```
Comment (AJCM-git) [inline Content.Shared/Climbing/BonkableComponent.cs:12]: Once you the next review is done, add this attribute `[Access(typeof(SharedBonkSystem)]`

File: Content.Server/Climbing/BonkSystem.cs [hunk 3]
```diff
@@ -0,0 +1,51 @@
+using Content.Server.Interaction;
+using Content.Server.Stunnable;
+using Content.Shared.CCVar;
+using Content.Shared.Climbing;
+using Content.Shared.Damage;
+using Content.Shared.DragDrop;
+using Robust.Server.GameObjects;
+using Robust.Shared.Configuration;
+
+namespace Content.Server.Climbing
```
Comment (AJCM-git) [inline Content.Server/Climbing/BonkSystem.cs:10]: Im pretty sure you can move this system to Content.Shared, also, use a file scoped namespace to reduce a level of identation

File: Content.Server/Climbing/BonkSystem.cs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        }
+
+        public bool TryBonk(EntityUid user, BonkableComponent bonkComponent)
+        {
+            if (!_cfg.GetCVar(CCVars.GameTableBonk))
+            {
+                // Not set to always bonk, try clumsy roll.
+                if (!_interactionSystem.TryRollClumsy(user, bonkComponent.BonkClumsyChance))
+                    return false;
+            }
```
Comment (AJCM-git) [inline Content.Server/Climbing/BonkSystem.cs:33]: resolve the bonkable component here

File: Content.Server/Climbing/ClimbSystem.cs [hunk 5]
```diff
@@ -289,7 +271,7 @@ private void StopClimb(EntityUid uid, ClimbingComponent? climbing = null, Fixtur
     /// <summary>
     ///     Checks if the user can vault the target
     /// </summary>
-    /// <param name="component">The component of the entity that is being vaulted</param>
+    /// <param name="component">The bonkComponent of the entity that is being vaulted</param>
```
Comment (AJCM-git) [inline Content.Server/Climbing/ClimbSystem.cs:274]: bonk component isnt used here

File: Content.Server/Climbing/ClimbSystem.cs [hunk 6]
```diff
@@ -324,7 +306,7 @@ private bool CanVault(ClimbableComponent component, EntityUid user, EntityUid ta
     /// <summary>
     ///     Checks if the user can vault the dragged entity onto the the target
     /// </summary>
-    /// <param name="component">The climbable component of the object being vaulted onto</param>
+    /// <param name="component">The climbable bonkComponent of the object being vaulted onto</param>
```
Comment (AJCM-git) [inline Content.Server/Climbing/ClimbSystem.cs:309]: same

File: Content.Server/Climbing/BonkSystem.cs [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+            _audioSystem.PlayPvs(bonkComponent.BonkSound, bonkComponent.Owner);
+            _stunSystem.TryParalyze(user, TimeSpan.FromSeconds(bonkComponent.BonkTime), true);
+
+            if (bonkComponent.BonkDamage is { } bonkDmg)
+                _damageableSystem.TryChangeDamage(user, bonkDmg, true, origin: user);
+
+            return true;
+        }
+
+        private void OnDragDrop(EntityUid user, BonkableComponent bonkComponent, DragDropEvent args)
```
Comment (AJCM-git) [inline Content.Server/Climbing/BonkSystem.cs:46]: does this still have the bug that if you drag drop someone else onto a table, you are the one getting bonked and not them?

File: Content.Shared/Climbing/BonkableComponent.cs [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    public sealed class BonkableComponent : Component
```
Comment (AJCM-git) [inline Content.Shared/Climbing/BonkableComponent.cs:13]: file scoped namespace to reduce indentation and once you move the system to shared add this attribute `[Access(typeof(BonkableSystem)] `

File: Content.Shared/Climbing/BonkableComponent.cs [hunk 9]
```diff
@@ -0,0 +1,42 @@
+using Content.Shared.Damage;
+using Robust.Shared.Audio;
+using Robust.Shared.GameStates;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
```
Comment (AJCM-git) [inline Content.Shared/Climbing/BonkableComponent.cs:8]: im pretty sure most of these arent used

File: Content.Server/Climbing/ClimbSystem.cs [hunk 10]
```diff
@@ -108,18 +106,25 @@ private void AddClimbableVerb(EntityUid uid, ClimbableComponent component, GetVe
...
-
+        if (!dragDrop)
+        {
+            BonkableComponent? bonkableComponent = null;
+            if (Resolve(climbable, ref bonkableComponent))
+            {
+                if (_bonkSystem.TryBonk(entityToMove, bonkableComponent))
+                    return;
+            }
+        }
```
Comment (AJCM-git) [inline Content.Server/Climbing/ClimbSystem.cs:126]: dont check for the drag drop,  the resolve should be done in TryBonk

File: Content.Server/Climbing/BonkSystem.cs [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+            if (!_cfg.GetCVar(CCVars.GameTableBonk))
+            {
+                // Not set to always bonk, try clumsy roll.
+                if (!_interactionSystem.TryRollClumsy(user, bonkComponent.BonkClumsyChance))
+                    return false;
+            }
+
+            // BONK!
+
+            _audioSystem.PlayPvs(bonkComponent.BonkSound, bonkComponent.Owner);
```
Comment (AJCM-git) [inline Content.Server/Climbing/BonkSystem.cs:37]: add a localized popup saying that you bonked your head or something

**Review by AJCM-git — CHANGES_REQUESTED** [review]

File: Content.Server/Climbing/BonkSystem.cs (hunk 7 above)
Comment (Telyonok) [reply Content.Server/Climbing/BonkSystem.cs:46]: No, fixed it

File: Content.Shared/Climbing/SharedBonkSystem.cs [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+            if (!_cfg.GetCVar(CCVars.GameTableBonk))
+            {
+                // Not set to always bonk, try clumsy roll.
+                if (!_interactionSystem.TryRollClumsy(user, bonkableComponent.BonkClumsyChance))
+                    return false;
+            }
+
+            // BONK!
+
+            _popupSystem.PopupEntity(BonkMessage, user, PopupType.Medium);
```
Comment (AJCM-git) [inline Content.Shared/Climbing/SharedBonkSystem.cs:42]: get the message with `Loc.GetString()` once you do the previous review and make two popups, one for yourself and one for others, here is one example from `DrinkSystem` (the identity part is important in this case)
```cs
                var targetName = Identity.Entity(uid, EntityManager);
                var userName = Identity.Entity(args.User, EntityManager);

                _popupSystem.PopupEntity(
                    Loc.GetString("drink-component-force-feed-success", ("user", userName), ("flavors", flavors)), uid, uid);

                _popupSystem.PopupEntity(
                    Loc.GetString("drink-component-force-feed-success-user", ("target", targetName)),
                    args.User, args.User);
```

**Review by AJCM-git — CHANGES_REQUESTED** [review]

File: Content.Shared/Climbing/BonkableComponent.cs [hunk 13]
```diff
@@ -0,0 +1,36 @@
+using Content.Shared.Damage;
+using Robust.Shared.Audio;
+using Robust.Shared.GameStates;
+
+namespace Content.Shared.Climbing;
+
+[RegisterComponent, NetworkedComponent, Access(typeof(SharedBonkSystem))]
+public sealed class BonkableComponent : Component
+{
```
Comment (metalgearsloth) [inline Content.Shared/Climbing/BonkableComponent.cs:9]: This needs documenting.

File: Content.Shared/Climbing/SharedBonkSystem.cs [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+using Content.Shared.Stunnable;
+using Content.Shared.CCVar;
+using Content.Shared.Damage;
+using Content.Shared.DragDrop;
+using Robust.Shared.Configuration;
+using Content.Shared.Popups;
+
+namespace Content.Shared.Climbing;
+
+public sealed class SharedBonkSystem : EntitySystem
```
Comment (metalgearsloth) [inline Content.Shared/Climbing/SharedBonkSystem.cs:11]: This system shouldn't have a "Shared" prefix.

**Review by AJCM-git — APPROVED** [review]

**Pull request merged**